File: py/lcs.py
def lcs_rec(x, i, m, y, j, n):

    if i == m or j==n:
        return 0

    if x[i]==y[j]:
        return lcs_rec(x,i+1,m, y, j+1, n)+1
    else:
        return max(lcs_rec(x, i+1, m, y, j, n), lcs_rec(x, i, m, y, j+1, n))


def lcs_dp(x, n, y, m):

    nn = n
    mm = m

    dp = [ [0 for i in range(m+1)] for j in range(n+1) ]

    for i in range(0, n+1):
        for j in range(m+1):

            if i==0 or j==0:
                dp[i][j]=0

            elif x[i-1]==y[j-1]:
                dp[i][j]= dp[i-1][j-1] +1
            else:
                dp[i][j] = max(dp[i-1][j], dp[i][j-1])


    lcs_len = dp[n][m]
    lcs_str = ['' for i in range(lcs_len)]
    lcs_len = lcs_len -1
    while (n>=0 and m>=0 and lcs_len>=0):

        if x[n-1] == y[m-1]:
            lcs_str[lcs_len]= x[n-1]
            n -=1
            m -=1
            lcs_len -=1

        elif dp[n-1][m] > dp[n][m-1]:
            n -=1

        else:
            m -=1

    print lcs_str


    return dp[nn][mm]


def lcs_opti(x, n, y, m):

    dp = [ [0 for i in range(m+1)] for j in range(2) ]

    idx = bool

    for i in range(0, n+1):
        idx = i & 1
        #print 'idx : ', idx
        for j in range(0, m+1):

            if i==0 or j==0:
                dp[idx][j]=0

            if x[i-1]==y[j-1]:
                dp[idx][j]= dp[1-idx][j-1]+1

            else:
                dp[idx][j]= max(dp[1-idx][j], dp[idx][j-1])

            #print dp

    return dp[idx][m]

if __name__=="__main__":

    #x = "AGGTAB" #"ABCDGH"
    x = "abcba"
    x_len = len(x)
    #y = "GXTXAYB" #"AEDFHR"
    y = "acab"
    y_len = len(y)

    #print 'LCS :: {}'.format(lcs_rec(x, 0, x_len, y, 0, y_len))


    #print 'LCS_DP :: {} '.format(lcs_dp(x, x_len, y, y_len))

    print 'LCS_OPTI :: {} '.format(lcs_opti(x, x_len, y, y_len))

File: cpp/practise/lcs.cpp
#include<bits/stdc++.h>

using namespace std;

int LCS(char X[], int xlen, int i, char Y[], int ylen, int j){

    if(i==xlen || j== ylen)
        return 0;

    if(X[i] == Y[j])
        return LCS(X, xlen, i+1, Y, ylen, j+1);
    else
        return max( LCS(X, xlen, i+1,Y, ylen, j), (X, xlen, i, Y, ylen, j+1) );
}

void LCSdp(char X[], int xlen, char Y[], int ylen){

    int lc[xlen+1][ylen+1];
    cout<<endl;
    for(int i=0; i<=xlen; i++){

        for(int j=0; j<=ylen; j++){

            if(i==0 || j==0)
                lc[i][j]=0;

            else if(X[i-1]==Y[j-1])
                lc[i][j] = lc[i-1][j-1]+1;
            else
                lc[i][j] = max(lc[i-1][j], lc[i][j-1]);

                cout<<lc[i][j]<<" ";
        }

        cout<<endl;
    }

//    return lc[xlen][ylen];

    int idx  = lc[xlen][ylen];
    char l[idx+1];
    l[idx] = '\0';

    int i= xlen; int j=ylen;
    while(i>0 && j>0){

            if(X[i-1]==Y[j-1]){
                l[idx-1] = X[ i-1];
                i--; j--; idx--;
            }
            else if(lc[i-1][j] > lc[i][j-1]){
                i--;
            }else{
                j--;
            }
    }

    cout<<"LCS = '"<<l<<"'"<<endl;
}

int LCSopti(char X[],int xlen, char Y[], int ylen){

    int L[2][ylen+1];
    bool idx;

    for(int i=0; i<=xlen; i++){
            idx = i&1;
        for(int j=0; j<=ylen ;j++){

                    if(i==0 || j==0)
                        L[idx][j]=0;
                    else if(X[i-1]==Y[j-1]){
                            L[idx][j] = L[1-idx][j-1]+1;
                    }else{
                            L[idx][j] = max(L[1-idx][j], L[idx][j-1]);
                    }

                    //cout<<L[idx][j]<<" ";
            }
            //cout<<endl;
    }

    return L[idx][ylen];
}

int main(){

    char X[] = "AGGTAB";
    //char X[] = "ABCDGH";
    int xlen = strlen(X);
    char Y[] = "GXTXAYB";
    //char Y[] = "AEDFHR";
    int ylen = strlen(Y);


    //cout<<"LCS  = "<<LCS(X, xlen, 0, Y, ylen, 0); O(2^n)
    //cout<<"LCS = "<<LCSdp(X, xlen, Y, ylen);      O(xlen * ylen)
    //LCSdp(X, xlen, Y, ylen);
    cout<<"LCS = "<<LCSopti(X, xlen, Y, ylen)<<endl;    //O(xlen * ylen)

    return 0;
}
